# Web Mining and Applied NLP (44-620)

## Requests, JSON, and NLP

### Student Name: Kate Huntsman

### GitHub Repo: https://github.com/katehuntsman/web-project-04-huntsman

This project involves fetching song lyrics from the public API `lyrics.ovh`, performing sentiment analysis on the lyrics using the `spaCy` library and its `spacytextblob` component, and analyzing the results. The aim is to explore the relationship between the sentiment polarity of lyrics and their thematic content.


In [20]:
import json
import pickle

import requests
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

print('All prereqs installed.')
!pip list

All prereqs installed.
Package                   Version
------------------------- --------------
annotated-types           0.7.0
anyio                     4.4.0
appnope                   0.1.4
argon2-cffi               23.1.0
argon2-cffi-bindings      21.2.0
arrow                     1.3.0
asttokens                 2.4.1
async-lru                 2.0.4
attrs                     24.2.0
babel                     2.16.0
beautifulsoup4            4.12.3
bleach                    6.1.0
blis                      1.0.1
build                     1.2.2
catalogue                 2.0.10
certifi                   2024.8.30
cffi                      1.17.1
charset-normalizer        3.4.0
click                     8.1.7
cloudpathlib              0.20.0
comm                      0.2.2
confection                0.1.5
cymem                     2.0.8
debugpy                   1.8.7
decorator                 5.1.1
defusedxml                0.7.1
en_core_web_sm            3.8.0
executing                 

1. The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

In [21]:
import requests
import json

# Fetch lyrics from the API
response = requests.get('https://api.lyrics.ovh/v1/They Might Be Giants/Birdhouse in your soul')
result = json.loads(response.text)

# Write the JSON data to a file
with open('lyrics.json', 'w') as json_file:
    json.dump(result, json_file, indent=4)

print("Lyrics saved to lyrics.json")

# Read the lyrics from the JSON file
with open('lyrics.json', 'r') as json_file:
    lyrics_data = json.load(json_file)

print(lyrics_data)


Lyrics saved to lyrics.json
{'lyrics': "I'm your only friend \nI'm not your only friend \nBut I'm a little glowing friend \nBut really I'm not actually your friend \nBut I am \n\n\nBlue canary in the outlet by the light switch \n\nWho watches over you \n\nMake a little birdhouse in your soul \n\nNot to put too fine a point on it \n\nSay I'm the only bee in your bonnet \n\nMake a little birdhouse in your soul \n\n\n\nI have a secret to tell \n\nFrom my electrical well \n\nIt's a simple message and I'm leaving out the whistles and bells \n\nSo the room must listen to me \n\nFilibuster vigilantly \n\nMy name is blue canary one note* spelled l-i-t-e \n\nMy story's infinite \n\nLike the Longines Symphonette it doesn't rest \n\n\n\nBlue canary in the outlet by the light switch \n\nWho watches over you \n\nMake a little birdhouse in your soul \n\nNot to put too fine a point on it \n\nSay I'm the only bee in your bonnet \n\nMake a little birdhouse in your soul \n\n\n\nI'm your only friend \n\n

2. Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

In [22]:
import json
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

# Register the SpacyTextBlob component
nlp.add_pipe("spacytextblob", last=True)

# Read the lyrics from the JSON file
with open('lyrics.json', 'r') as json_file:
    lyrics_data = json.load(json_file)

# Extract the lyrics from the dictionary
lyrics = lyrics_data.get('lyrics', '')

# Perform sentiment analysis
doc = nlp(lyrics)
polarity_score = doc._.blob.polarity  # Access polarity from the TextBlob object

# Print the lyrics and the polarity score
print("Lyrics:\n", lyrics)
print("\nPolarity Score:", polarity_score)

# Comment on the sentiment
# Based on the polarity score, if it is greater than 0, the lyrics have a more positive connotation.
# If it is less than 0, they have a more negative connotation. 
# A score close to 0 suggests a neutral sentiment.
if polarity_score > 0:
    print("The lyrics have a more positive connotation.")
elif polarity_score < 0:
    print("The lyrics have a more negative connotation.")
else:
    print("The lyrics have a neutral connotation.")

Lyrics:
 I'm your only friend 
I'm not your only friend 
But I'm a little glowing friend 
But really I'm not actually your friend 
But I am 


Blue canary in the outlet by the light switch 

Who watches over you 

Make a little birdhouse in your soul 

Not to put too fine a point on it 

Say I'm the only bee in your bonnet 

Make a little birdhouse in your soul 



I have a secret to tell 

From my electrical well 

It's a simple message and I'm leaving out the whistles and bells 

So the room must listen to me 

Filibuster vigilantly 

My name is blue canary one note* spelled l-i-t-e 

My story's infinite 

Like the Longines Symphonette it doesn't rest 



Blue canary in the outlet by the light switch 

Who watches over you 

Make a little birdhouse in your soul 

Not to put too fine a point on it 

Say I'm the only bee in your bonnet 

Make a little birdhouse in your soul 



I'm your only friend 

I'm not your only friend 

But I'm a little glowing friend 

But really I'm not actual

3. Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.

In [23]:
import requests
import json

def fetch_and_save_lyrics(artist, song, filename):
    """Fetch lyrics from lyrics.ovh API and save to a file."""
    url = f'https://api.lyrics.ovh/v1/{artist}/{song}'
    response = requests.get(url)
    
    if response.status_code == 200:
        lyrics_data = response.json()
        # Write the results to the specified file
        with open(filename, 'w') as json_file:
            json.dump(lyrics_data, json_file, indent=4)
        print(f"Lyrics for '{song}' by {artist} saved to '{filename}'.")
    else:
        print(f"Failed to fetch lyrics for '{song}' by {artist}. Status Code: {response.status_code}")

# Test the function with four songs
fetch_and_save_lyrics("They Might Be Giants", "Birdhouse in Your Soul", "birdhouse_lyrics.json")
fetch_and_save_lyrics("The Beatles", "Hey Jude", "hey_jude_lyrics.json")
fetch_and_save_lyrics("Taylor Swift", "Love Story", "love_story_lyrics.json")
fetch_and_save_lyrics("Adele", "Hello", "hello_lyrics.json")

Lyrics for 'Birdhouse in Your Soul' by They Might Be Giants saved to 'birdhouse_lyrics.json'.
Lyrics for 'Hey Jude' by The Beatles saved to 'hey_jude_lyrics.json'.
Lyrics for 'Love Story' by Taylor Swift saved to 'love_story_lyrics.json'.
Lyrics for 'Hello' by Adele saved to 'hello_lyrics.json'.


4. Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

In [24]:
import json
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

# Load the spaCy model and add SpacyTextBlob
nlp = spacy.load('en_core_web_sm')
nlp.add_pipe("spacytextblob", last=True)

def analyze_lyrics_sentiment(filename):
    """Load lyrics from a file and return the polarity score."""
    with open(filename, 'r') as json_file:
        lyrics_data = json.load(json_file)

    # Extract the lyrics
    lyrics = lyrics_data.get('lyrics', '')
    
    # Perform sentiment analysis
    doc = nlp(lyrics)
    polarity_score = doc._.blob.polarity
    return polarity_score

# Analyze and print the polarity scores for the three files
songs_files = [
    ("birdhouse_lyrics.json", "Birdhouse in Your Soul"),
    ("hey_jude_lyrics.json", "Hey Jude"),
    ("love_story_lyrics.json", "Love Story"),
]

for filename, song in songs_files:
    polarity = analyze_lyrics_sentiment(filename)
    print(f"The polarity score for '{song}' is: {polarity}")

The polarity score for 'Birdhouse in Your Soul' is: 0.04505208333333333
The polarity score for 'Hey Jude' is: 0.13194444444444445
The polarity score for 'Love Story' is: 0.049999999999999996


## Analysis:
### Birdhouse in Your Soul
Polarity Score: 0.045
Analysis: This score is slightly positive but quite close to neutral (0.0). Given the whimsical and playful nature of the song, one might expect a higher score. This suggests that while the lyrics contain positive elements, the sentiment analysis may not fully capture the lighthearted and nostalgic essence of the song.

### Hey Jude
Polarity Score: 0.132
Analysis: This score indicates a moderately positive sentiment. It aligns with the song's themes of encouragement and reassurance. The message of making things better and supporting someone in tough times resonates well with a positive score. However, it might still be lower than expected for such an uplifting anthem, reflecting possible complexities in the lyrics that aren't fully captured.

### Love Story
Polarity Score: 0.050
Analysis: Similar to "Birdhouse in Your Soul," this score is just above neutral. Given the romantic and hopeful narrative of the song, one would anticipate a more robust positive score. The slight positivity suggests that while the song conveys love and hope, the analysis may miss some of the depth and emotional richness in the lyrics.

### Overall Reflection
General Trends: The scores suggest that all three songs are perceived as somewhat positive, but not strongly so. This could indicate that the sentiment analysis tool may struggle with the nuanced emotions present in these lyrics.

Limitations of Sentiment Analysis: The subtlety in lyrical themes, like nostalgia, irony, or complex emotions, might not be well represented by the polarity score. This limitation highlights the challenge of relying solely on automated sentiment analysis for a complete understanding of song lyrics.